mentor/messages.py

```diff
@@ -0,0 +1,370 @@
+        except:
+            return False
+
+    # only called if k not found in normal places
+    def __getattr__(self, k):
+        """ Gets key if it exists, otherwise throws AttributeError.
+        """
+        try:
+            # Throws exception if not in prototype chain
+            return object.__getattribute__(self, k)
+        except AttributeError:
+            try:
+                return self[k]
+            except KeyError:
+                raise AttributeError(k)
+
+    def __setattr__(self, k, v):
+        """
+        """
+        try:
+            # Throws exception if not in prototype chain
+            object.__getattribute__(self, k)
+        except AttributeError:
+            try:
+                self[k] = v
+            except:
+                raise AttributeError(k)
+        else:
+            object.__setattr__(self, k, v)
+
+    def __delattr__(self, k):
+        """ Deletes attribute k if it exists, otherwise deletes key k. A KeyError
+            raised by deleting the key--such as when the key is missing--will
+            propagate as an AttributeError instead.
+        """
+        try:
+            # Throws exception if not in prototype chain
+            object.__getattribute__(self, k)
+        except AttributeError:
+            try:
+                del self[k]
+            except KeyError:
+                raise AttributeError(k)
+        else:
+            object.__delattr__(self, k)
+
+
+    def __repr__(self):
+        """ Invertible* string-form of a Message.
+        """
+        keys = list(iterkeys(self))
+        keys.sort()
+        args = ', '.join(['%s=%r' % (key, self[key]) for key in keys])
+        return '%s(%s)' % (self.__class__.__name__, args)
+
+    def __dir__(self):
+        return self.keys()
+
+def Cpus(value):
+    return Message({'name': 'cpus', 'role': '*', 'scalar': Message({'value':value}), 'type': 'SCALAR'})
+
+def Mem(value):
+    return Message({'name': 'mem', 'role': '*', 'scalar':  Message({'value':value}), 'type': 'SCALAR'})
+
+def Disk(value):
+    return Message({'name': 'disk', 'role': '*', 'scalar': Message({'value':value}), 'type': 'SCALAR'})
+
+def Ports(begin,end):
+    return Message({'name': 'disk', 'role': '*', 'ranges': Message({'range':[Message({'begin': begin, 'end': end})]}), 'type': 'SCALAR'})
+
+
+class ResourceMixin(object):
+
+    zeroes = {'cpus': 0.0, 'mem': 0.0, 'disk': 0.0, 'ports': [{"begin":0.0,"end":0.0}]} # gpu
+
+    @staticmethod
+    def flatten(message):
+        flattened = {}
+        for r in message['resources']:
+            if r["type"]=="RANGES":
+                flattened[r['name']] = r['ranges']['range']
+            else:
+                flattened[r['name']] = r['scalar']['value']
+        return flattened
+
+    def apply(self, op, other):
+        a = self.flatten(self)
+        b = self.flatten(other)
+        out = {}
+        for k, v in self.zeroes.items():
+            if k=="ports":
```

Comment:
With supporting ports it could be cleaner to create scalar and range types and convert the resources when initializing the ResourceMixin.